hvac/v1/__init__.py

```diff
@@ -598,6 +602,63 @@ def auth_userpass(self, username, password, mount_point='userpass', use_token=Tr
 
         return self.auth('/v1/auth/{0}/login/{1}'.format(mount_point, username), json=params, use_token=use_token)
 
+    def auth_aws_iam(self, access_key, secret_key, session_token=None, header_value=None, mount_point='aws', role='', use_token=True):
+        """
+        POST /auth/<mount point>/login
+        """
+        timestamp = datetime.utcnow().strftime('%Y%m%dT%H%M%SZ')
+
+        method = 'POST'
+        url = 'https://sts.amazonaws.com/'
+        body = 'Action=GetCallerIdentity&Version=2011-06-15'
+        headers = {
+            'Content-Length': [str(len(body))],
+            'Content-Type': ['application/x-www-form-urlencoded; charset=utf-8'],
+            'Host': ['sts.amazonaws.com'],
+            'X-Amz-Date': [timestamp]
+        }
+
+        if session_token:
+            headers['X-Amz-Security-Token'] = [session_token]
+        if header_value:
+            headers['X-Vault-AWS-IAM-Server-ID'] = [header_value]
+
+        # https://docs.aws.amazon.com/general/latest/gr/sigv4-create-canonical-request.html
+        canonical_headers = ''.join('{0}:{1}\n'.format(k.lower(), headers[k][0]) for k in sorted(headers))
+        signed_headers = ';'.join(k.lower() for k in sorted(headers))
+        payload_hash = sha256(body.encode('utf-8')).hexdigest()
+        canonical_request = '\n'.join([method, '/', '', canonical_headers, signed_headers, payload_hash])
+
+        # https://docs.aws.amazon.com/general/latest/gr/sigv4-create-string-to-sign.html
+        algorithm = 'AWS4-HMAC-SHA256'
+        credential_scope = '/'.join([timestamp[0:8], 'us-east-1', 'sts', 'aws4_request'])
+        canonical_request_hash = sha256(canonical_request.encode('utf-8')).hexdigest()
+        string_to_sign = '\n'.join([algorithm, timestamp, credential_scope, canonical_request_hash])
+
+        # https://docs.aws.amazon.com/general/latest/gr/sigv4-calculate-signature.html
+        key = 'AWS4{0}'.format(secret_key).encode('utf-8')
```

Comment:
The copious amount of `encode('utf-8')` calls give me some pause. I'm no expert at understanding all the implications that come with character encodings in python 2/3, but I have typically found if I'm having to fiddle with encodings this much, there is often an easier way of ending up with the same result. What was the motivation for including these?